velox/common/base/tests/AsyncSourceTest.cpp

```diff
@@ -151,3 +151,97 @@ TEST(AsyncSourceTest, errorsWithThreads) {
   // gizmo is sure to get an error.
   EXPECT_LT(0, numErrors);
 }
+
+class DataCounter {
+ public:
+  DataCounter() {
+    objectNumber_ = ++numCreatedDataCounters_;
+  }
+
+  ~DataCounter() {
+    ++numDeletedDataCounters_;
+  }
+
+  static uint64_t numCreatedDataCounters() {
+    return numCreatedDataCounters_;
+  }
+
+  static uint64_t numDeletedDataCounters() {
+    return numDeletedDataCounters_;
+  }
+
+  static void reset() {
+    numCreatedDataCounters_ = 0;
+    numDeletedDataCounters_ = 0;
+  }
+
+  uint64_t objectNumber() const {
+    return objectNumber_;
+  }
+
+ private:
+  static std::atomic<uint64_t> numCreatedDataCounters_;
+  static std::atomic<uint64_t> numDeletedDataCounters_;
+
+  uint64_t objectNumber_{0};
+};
+
+std::atomic<uint64_t> DataCounter::numCreatedDataCounters_ = 0;
+
+std::atomic<uint64_t> DataCounter::numDeletedDataCounters_ = 0;
+
+TEST(AsyncSourceTest, close) {
+  // If 'prepare()' is not executed within the thread pool, invoking 'close()'
+  // will set 'make_' to nullptr. The deletion of 'dateCounter' is used as a
+  // verification for this behavior.
+  auto dateCounter = std::make_shared<DataCounter>();
+  AsyncSource<uint64_t> countAsyncSource([dateCounter]() {
+    return std::make_unique<uint64_t>(dateCounter->objectNumber());
+  });
+  dateCounter.reset();
+  EXPECT_EQ(DataCounter::numCreatedDataCounters(), 1);
+  EXPECT_EQ(DataCounter::numDeletedDataCounters(), 0);
+
+  countAsyncSource.close();
+  EXPECT_EQ(DataCounter::numCreatedDataCounters(), 1);
+  EXPECT_EQ(DataCounter::numDeletedDataCounters(), 1);
+  DataCounter::reset();
+
+  // If 'prepare()' is executed within the thread pool but 'move()' is not
+  // invoked, invoking 'close()' will set 'item_' to nullptr. The deletion of
+  // 'dateCounter' is used as a verification for this behavior.
+  auto asyncSource = std::make_shared<AsyncSource<DataCounter>>([]() {
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
+    return std::make_unique<DataCounter>();
+  });
+  auto thread = std::thread([&asyncSource] { asyncSource->prepare(); });
+  thread.join();
+  EXPECT_EQ(DataCounter::numCreatedDataCounters(), 1);
+  EXPECT_EQ(DataCounter::numDeletedDataCounters(), 0);
+
+  asyncSource->close();
+  EXPECT_EQ(DataCounter::numCreatedDataCounters(), 1);
+  EXPECT_EQ(DataCounter::numDeletedDataCounters(), 1);
+  DataCounter::reset();
+
+  // If 'prepare()' is currently being executed within the thread pool,
+  // 'close()' should wait for the completion of 'prepare()' and set 'item_' to
+  // nullptr.
+  std::atomic<bool> started = false;
+  auto sleepAsyncSource =
+      std::make_shared<AsyncSource<DataCounter>>([&started]() {
+        started = true;
+        std::this_thread::sleep_for(std::chrono::milliseconds(10));
```

Comment:
You can use `folly::Baton::wait` to pause & resume the thread instead of doing sleep here and below

Reply:
Thanks, updated.